csrc/device_lower/utils.cpp

```diff
@@ -1013,6 +1016,775 @@ bool predicateAtEnd(ForLoop* loop) {
+Val* extent(const PartOf<Projection>& part) {
+  return part.selected_extent;
+}
+
+Val* extent(const Composition<Projection>& comp) {
+  return std::accumulate(
+      comp.begin(),
+      comp.end(),
+      static_cast<Val*>(nullptr),
+      [](Val* acc, const auto& g) {
+        return SimplifyingIrBuilder::mulExpr(acc, extent(g));
+      });
+}
+
+Val* extent(const std::monostate&) {
+  NVF_ERROR(false, "Cannot get extent of std::monostate");
+}
+
+Val* extent(const Projection& proj) {
+  return Projection::dispatch(
+      [&](const auto& proj) { return extent(proj); }, proj);
+}
+
+// Simplify the abstract syntax tree so that it is easier to be pattern
+// matched.
+Projection simplify(Projection proj);
+
+Projection simplify(const ValGroup& group) {
+  return group;
+}
+
+PartOf<Projection> cancelCommonFactors(const PartOf<Projection>& part) {
+  // If `what` is a composition and inner_extent is a multiple of the extent of
+  // the last items in `what`, we can simplify the inner_extent and `what` by
+  // canceling the last items in `what`.
+  //
+  // Example:
+  // PartOf{what=[5,3,2], inner_extent=42} => PartOf{what=[5], inner_extent=7}
+  if (!part.what->is<Composition>()) {
+    return part;
+  }
+  auto dq = part.what->as<Composition>();
+
+  Val* new_inner_extent = part.inner_extent;
+  if (new_inner_extent == nullptr) {
+    return part;
+  }
+
+  while (!dq.empty() &&
+         simplifyExpr(
+             IrBuilder::isDivisibleExpr(new_inner_extent, extent(dq.back())))
+             ->isTrue()) {
+    new_inner_extent =
+        simplifyExpr(IrBuilder::divExpr(new_inner_extent, extent(dq.back())));
+    dq.pop_back();
+  }
+  if (new_inner_extent->isOne()) {
+    new_inner_extent = nullptr;
+  }
+  NVF_ERROR(!dq.empty());
+  if (dq.size() == 1) {
+    return PartOf<Projection>{
+        std::make_shared<Projection>(dq.front()),
+        new_inner_extent,
+        part.selected_extent};
+  }
+  return PartOf<Projection>{
+      std::make_shared<Projection>(std::move(dq)),
+      new_inner_extent,
+      part.selected_extent};
+}
+
+PartOf<Projection> trimRedundant(const PartOf<Projection>& part) {
+  // If part.what is a composition then we only keep the minimum number of
+  // items in `what` that is sufficient to represent the selected_extent *
+  // inner_extent.
+  //
+  // Example:
+  // PartOf{what=[7, 3, 5, 2], inner_extent=3, selected_extent=5} =>
+  //   PartOf{what=[3, 5, 2], inner_extent=3, selected_extent=5}
+  if (!part.what->is<Composition>()) {
+    return part;
+  }
+  auto dq = part.what->as<Composition>();
+
+  Val* required_extent =
+      SimplifyingIrBuilder::mulExpr(part.selected_extent, part.inner_extent);
+
+  Val* what_extent = nullptr;
+  int64_t count = 0;
+  while (count < dq.size()) {
+    count++;
+    const auto& item = dq.at(dq.size() - count);
+    what_extent = SimplifyingIrBuilder::mulExpr(what_extent, extent(item));
+    if (simplifyExpr(IrBuilder::isDivisibleExpr(what_extent, required_extent))
```

Comment:
Can this be just `what_extent >= required_extent`?

Reply:
I believe no, for similar reason as https://github.com/NVIDIA/Fuser/pull/415 where we changed `>=` to `gcd` in vectorization helper.

Reply:
Thanks. Can you please add this as a comment?

Reply:
Added a "Proof of correctness" for `trimRedundant`.